beginner_source/onnx/export_control_flow_model_to_onnx_tutorial.py

```diff
@@ -0,0 +1,104 @@
+# -*- coding: utf-8 -*-
+"""
+**Introduction to ONNX** ||
+`Exporting a PyTorch model to ONNX <export_simple_model_to_onnx_tutorial.html>`_ ||
+`Extending the ONNX Registry <onnx_registry_tutorial.html>`_
+
+Export a control flow model to ONNX
+==========================================
+
+**Author**: `Xavier Dupré <https://github.com/xadupre>`_.
+
+Tests cannot be exported into ONNX unless they refactored
+to use :func:`torch.cond`. Let's start with a simple model
+implementing a test.
+"""
+
+from onnx.printer import to_text
+import torch
+
+class ForwardWithControlFlowTest(torch.nn.Module):
+    def forward(self, x):
+        if x.sum():
+            return x * 2
+        return -x
+
+
+class ModelWithControlFlowTest(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.mlp = torch.nn.Sequential(
+            torch.nn.Linear(3, 2),
+            torch.nn.Linear(2, 1),
+            ForwardWithControlFlowTest(),
+        )
+
+    def forward(self, x):
+        out = self.mlp(x)
+        return out
+
+
+model = ModelWithControlFlowTest()
+
+# %%
+# Let's check it runs.
+x = torch.randn(3)
+model(x)
+
+# %%
+# As expected, it does not export.
+try:
+    torch.export.export(model, (x,))
+    raise AssertionError("This export should failed unless pytorch now supports this model.")
+except Exception as e:
+    print(e)
+
+# %%
+# It does export with :func:`torch.onnx.export` because
+# it uses JIT to trace the execution.
+# But the model is not exactly the same as the initial model.
+ep = torch.onnx.export(model, (x,), dynamo=True)
+print(to_text(ep.model_proto))
+
+
+# %%
+# Suggested Patch
+# +++++++++++++++
+#
+# Let's avoid the graph break by replacing the forward.
+
+
+def new_forward(x):
+    def identity2(x):
+        return x * 2
+
+    def neg(x):
+        return -x
+
+    return torch.cond(x.sum() > 0, identity2, neg, (x,))
+
+
+print("the list of submodules")
+for name, mod in model.named_modules():
+    print(name, type(mod))
+    if isinstance(mod, ForwardWithControlFlowTest):
+        mod.forward = new_forward
+
+# %%
+# Let's see what the fx graph looks like.
+
+print(torch.export.export(model, (x,)).graph)
+
+# %%
+# Let's export again.
+
+ep = torch.onnx.export(model, (x,), dynamo=True)
+print(to_text(ep.model_proto))
+
```

Comment:
We should include the result of the print here I think

Reply:
Like this where we include the print output: https://github.com/pytorch/tutorials/pull/3223/files#diff-974479b5e9f94c102ca2630bcc39ee4e63d28f220dc1cea6198fc0083396a2bcR107-R135

Reply:
I looked at other examples in the tutorials, output was not included. The generation should take care of that otherwise we would have to update them everytime pytorch is released. I guess they do something very similar to what sphinx-gallery does.

Reply:
Yes, it should be printed in the page. The code will be executed.

Reply:
I see. Strange I was looking at the registry page and there was no print out

Reply:
I was looking at https://pytorch.org/tutorials/beginner/onnx/onnx_registry_tutorial.html

Reply:
I'd rather remove every outputs, it is a bug if it does not appear. pytorch, is using sphinx-gallery. There should be warning on the output. I was looking at conf.py to see some explanations. No luck so far. We should ask meta if something does not work.

Reply:
That's fair